python/tk_hiero_export/sg_shot_processor.py

```diff
@@ -246,35 +300,283 @@ def startProcessing(self, exportItems):
+        # they can use during execution
+        cut_related_tasks = []
 
-        tasks = []
+        # iterate over the tasks groups to be executed
         for taskGroup in self._submission.children():
+
+            # placeholders for the tasks we want to pre-process
+            (transcode_task, shot_updater_task) = (None, None)
+
+            # look at all the tasks in the group and identify the shot updater
+            # and transcode tasks.
             for task in taskGroup.children():
+
+                # shot updater
                 if isinstance(task, ShotgunShotUpdater):
                     if task.isCollated():
                         # For collating sequences, skip tasks that are not hero
                         if task.isHero():
-                            tasks.append(task)
+                            shot_updater_task = task
                     else:
                         # For non-collating sequences, add every task
-                        tasks.append(task)
+                        shot_updater_task = task
+
+                # transcode
+                elif isinstance(task, ShotgunTranscodeExporter):
+                    transcode_task = task
+
+            # add the pair of associated tasks to the list of cut related tasks
+            cut_related_tasks.append((shot_updater_task, transcode_task))
+
+        # sort the tasks based on their position in the timeline. this gives
+        # us the cut order.
+        cut_related_tasks.sort(key=lambda tb: tb[0]._item.timelineIn())
+
+        # go ahead and populate the shot updater tasks with the cut order. this
+        # is used to set the cut order on the Shot as it is created/updated.
+        for i in range(0, len(cut_related_tasks)):
+            (shot_updater_task, transcode_task) = cut_related_tasks[i]
+
+            # Cut order is 1-based
+            shot_updater_task._cut_order = i + 1
+
+        # if you're wondering why we looped over the tasks above only to bail
+        # out here if cuts support isn't available for the site, it's to
+        # maintain backward compatibility for updating the Shot entities with
+        # cut data which relies on setting `_cut_order` on those updater tasks.
+        # The we also use the above loops to get the tasks in cut order which
+        # will be used in `_processCuts` if the site has cut support.
+
+        if not _cutsSupported(self.app.shotgun):
+            # cuts not supported. all done here
+            return
+
+        # ---- at this point, we have the cut related tasks in order.
+
+        # Create the Cut and CutItem entries for this submission. This will be
+        # done in a batch call, so show the busy popup while this is going on.
+        # Otherwise, it may look like Hiero is hanging.
+        self.app.engine.show_busy(
+            "Preprocessing Sequence",
+            "Creating Cut and Cut Items in Shotgun ..."
+        )
```

Comment:
Is there any way to defer this work to the tasks themselves so that we don't block here?

Would it work to have each shot updater create the CutItems and have the first one create the Cut, or some other way to tie this together without pausing at this point?